### ISSUE TITLE
Custom requests should be allowed to provide an implementation of on_success

### ISSUE THREAD

**Issue by cjh1** [celery/celery#4526]
## Checklist

- [x] I have included the output of ``celery -A proj report`` in the issue.
      (if you are not able to do this, then at least specify the Celery
       version affected).
- [x] I have verified that the issue exists against the `master` branch of Celery.

## Steps to reproduce
Write custom request with a ```on_success(...)``` method.
## Expected behavior
The method would be called when the request is complete.
## Actual behavior
The method is not called.

See #3977 for more details. The fix is probably to add a call to super [here](https://github.com/celery/celery/blob/master/celery/worker/request.py#L539).

**Comment by mvaled** [comment]
I would go the other way: Making the custom Request the base for the actual Request class.

**Comment by cjh1** [comment]
@mvaled Not sure I understand, isn't that what is happend [here](https://github.com/celery/celery/blob/master/celery/worker/request.py#L513)?

**Comment by mvaled** [comment]
@cjh1 If I recall correctly, the `base` there is the custom request, so the custom request is not in control of what to do.  Nevertheless, I don't have a real-life case that would require me to hack the `on_success`.  So, calling super may be just fine.

**Comment by cjh1** [comment]
@mvaled I have pushed an update that removes the implementation in ```create_request_cls``` which is essentially a duplicate of the implementation in ```Request```, thus an implementation in a custom request would now get called. The PR is #4527

**Comment by mvaled** [comment]
@cjh1 That's right.  So I would suggest a note in the docs, advising to call super in the custom request.

**Comment by mrezzamoradi** [comment]
We have a real issue with this after upgrading from 3.1.25 to 5.1.2 (yes it was huge). we had a code to patch the `Request.on_success` method to calculate the overall execution time of a successful task using `Request.time_start`. After the upgrade, we tried to use custom request but the `create_request_cls` helper overrides our custom request `on_success` method.
I couldn't reason about the existence of `create_request_cls` helper. it is a bad programming practice anyway, as it:
- causes a maintainer to keep the both pairs of `on_success` and `execute_using_pool` methods in sync (in the request.py).
- users expect their custom Request.on_success to take effect

**Comment by mrezzamoradi** [comment]
I understand it's a bit complicated to remove `create_request_cls`. is it possible to introduce a flag like `cusome=True` to the Request class and let `create_request_cls` return a class based on that? or alternatively use `mro_lookup` in `create_request_cls` to see if the request class has a custom `on_success/execute_using_pool`
